src/wallet/wallet.cpp

```diff
@@ -3900,6 +3941,42 @@ int CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool fUpdate)
             pindex = chainActive.Next(pindex);
         }
 
+        // Attempt to rewind the orchard wallet to the rescan point if the wallet has any
+        // checkpoints. Note data will be restored by the calls to AddToWalletIfInvolvingMe,
+        // and then the call to `ChainTipAdded` that later occurs for each block will restore
+        // the witness data that is being removed in the rewind here.
+        auto nu5_height = chainParams.GetConsensus().GetActivationHeight(Consensus::UPGRADE_NU5);
+        bool performOrchardWalletUpdates{false};
+        if (orchardWallet.IsCheckpointed()) {
+            // We have a checkpoint, so attempt to rewind the Orchard wallet at most as
+            // far as the NU5 activation block.
+            // If there's no activation height, we shouldn't have a checkpoint already,
+            // and this is a programming error.
+            assert(nu5_height.has_value());
+            // Only attempt to perform scans for Orchard during wallet initialization,
+            // since we do not support Orchard key import.
+            if (isInitScan) {
+                int rewindHeight = std::max(nu5_height.value(), pindex->nHeight);
+                uint32_t blocksRewound{0};
+                if (orchardWallet.Rewind(rewindHeight, blocksRewound)) {
+                    // rewind was successful or a no-op, so perform Orchard wallet updates
+                    LogPrintf("Rewound Orchard wallet by %d blocks.", blocksRewound);
+                    performOrchardWalletUpdates = true;
+                } else {
+                    // Orchard witnesses will not be able to be correctly updated, because we
+                    // can't rewind far enough. This is an unrecoverable failure; it means that we
+                    // can't get back to a valid wallet state without resetting the wallet all
+                    // the way back to NU5 activation.
+                    throw std::runtime_error("CWallet::ScanForWalletTransactions(): Orchard wallet is out of sync. Please restart your node with -rescan.");
+                }
+            }
+        } else if (isInitScan && pindex->nHeight < nu5_height) {
+            // If it's the initial scan and we're starting below the nu5 activation
+            // height, we're effectively rescanning from genesis and so it's safe
+            // to update the note commitment tree as we progress.
+            performOrchardWalletUpdates = true;
```

Comment:
This assumes the tree has been reset prior to calling `CWallet::ScanForWalletTransactions`. Why is that reset occurring in `InitLoadWallet` instead of right here?

Reply:
`Reset` is only called when an explicit `-rescan` is requested.